.github/actions/cluster/lxd-cluster.sh

```diff
@@ -0,0 +1,292 @@
+                fi
+
+                lxc launch "${INSTANCE_IMAGE}" "${instance}" \
+                        --storage "${STORAGE_POOL}" \
+                        --network "${NETWORK_NAME}" \
+                        -c limits.cpu=4 \
+                        -c limits.memory=4GiB \
+                        $args
+        done
+
+        # Wait for instances to become ready.
+        for i in $(seq 1 "${CLUSTER_SIZE}"); do
+                instance="${INSTANCE}-${i}"
+                waitInstance "${instance}"
+                lxc exec "${instance}" -- systemctl is-system-running --wait
+        done
+
+        # Install LXD on VMs.
+        for i in $(seq 1 "${CLUSTER_SIZE}"); do
+                instance="${INSTANCE}-${i}"
+
+                echo "Preparing instance ${instance} ..."
+
+                # Install snap daemon.
+                lxc exec "${instance}" --env=DEBIAN_FRONTEND=noninteractive -- bash -c "\
+                        apt-get -qq -y install snapd" > /dev/null
+
+                # Install LXD snap.
+                lxc exec "${instance}" -- bash -c "snap install lxd --channel ${VERSION_LXD} || sudo snap refresh lxd --channel ${VERSION_LXD}" > /dev/null
+        done
+
+        echo "Cluster instances created."
+        lxc list
+}
+
+# configure_lxd configures LXD cluster.
+configure_lxd() {
+        echo "Creating LXD cluster ..."
+
+        # Create LXD cluster.
+        for i in $(seq 1 "${CLUSTER_SIZE}"); do
+                instance="${INSTANCE}-${i}"
+
+                isClustered=$(lxc exec "${instance}" -- bash -c "lxc cluster list 2> /dev/null" || true)
+                if [ "${isClustered}" ]; then
+                        continue
+                fi
+
+                # Get IPv4 of the instance.
+                ipv4=$(instanceIPv4 "${instance}")
+
+                # On the leader instance, just enable clustering and continue.
+                if [ "${instance}" = "${LEADER}" ]; then
+                        lxc exec "${instance}" -- lxc config set core.https_address "${ipv4}"
+                        lxc exec "${instance}" -- lxc cluster enable "${instance}"
+                        continue
+                fi
+
+                # Create and extract token for a new cluster member.
+                token=$(lxc exec "${LEADER}" -- lxc cluster add -q "${instance}")
+                if [ "${token}" = "" ]; then
+                        echo "Error: Failed retrieveing join token for instance ${instance}"
+                        exit 1
+                fi
+
+                # Create temporary preseed config file.
+                config="/tmp/${instance}.cluster.cfg"
+                echo -e "\
+cluster:
+  enabled: true
+  server_address: ${ipv4}
+  cluster_token: ${token}
+" > "${config}"
+
+                # Apply the config and remove the temporary file.
+                cat "${config}" | lxc exec "${instance}" -- bash -c "lxd init --preseed"
+                rm "${config}"
```

Comment:
This doesn't require a temporary file. See https://github.com/canonical/lxd-ci/blob/main/tests/cluster#L70-L79 for inspiration.